active/0020-node-evacuation.md

```diff
@@ -0,0 +1,399 @@
+
+When being enabled, Eviction Agent enables prohibiting hooks: `client.connect` and `client.connack`;
+
+The internal API is the following:
+
+```erlang
+ok = emqx_eviction_agent:enable(node_evacuation, <<"host1:port1 host2:port2">>).
+```
+
+##### Enabled (`enabled`)
+
+When enabled, Eviction Agent has two working hooks. Their logic is:
+* `client.connect` hook returns `CONNACK_SERVER` (for MQTT 3) or `RC_USE_ANOTHER_SERVER` (for MQTT 5) error.
+* `client.connack` sets `'Server-Reference'` property to the specified servers `host1:port1 host2:port2 ...` for MQTT 5
+clients acked with `RC_USE_ANOTHER_SERVER`. `'Server-Reference'` property is not set if servers are not specified or
+for MQTT 3 clients.
+
+HTTP responses are the following:
+
+Request:
+```
+GET /api/v4/node_eviction/status
+```
+
+Response:
+```
+HTTP/1.1 200 OK
+
+{
+    "status": "enabled",
+    "stats": {
+        "current_connected": 25,
+        "current_sessions": 50
+    }
+}
+```
+
+Request:
+```
+GET /api/v4/node_eviction/heath_check
+```
+
+Response:
+```
+HTTP/1.1 503 OK
+```
+
+The internal API behavior is the following:
+
+```erlang
+{enabled, Kind,
+   #{connections := ConnectionCount,
+     sessions := TotalCount}} = emqx_eviction_agent:status(),
+ConnEvictCount = 10,
+ok = emqx_eviction_agent:evict_connections(ConnEvictCount),
+SessionEvictCount = 10,
+ok = emqx_eviction_agent:evict_sessions(ToNode, SessionEvictCount).
+```
+
+##### Disabling
+On disable either being enabled, Eviction Agent disables installed hooks.
+
+
+### Evacuation Coordinator (**without** enforced session evacuation)
+
+The Evacuation Coordinator is a stateful process that governs evacuation, a kind of Rebalance Coordinator.
+
+The Evacuation Coordinator has two states: `evicting` and `disabled`. It preserves its settings in a local file,
+so it restores its state even after node restart.
+
+Its logic is relatively straightforward.
+
+#### Algorithm
+
+##### Starting
+
+At the start, the Evacuation Coordinator reads the settings from the local file.
+* If evacuation is enabled Evacuation Coordinator tries to pass to the `evicting` state. It enables the local Eviction Agent.
+    * If the local Eviction Agent is successfully enabled, then the Evacuation Coordinator passed to the `evicting` state.
+    * If the local Eviction Agent is already enabled with a different `kind` (i.e., is occupied by some other coordinator, like rebalancing), then the Evacuation Coordinator passes to the `disabled` state.
+* The Evacuation Coordinator passes to the `disabled` state if evacuation is not enabled.
+
+##### `disabled` state
+This is the default no-op state.
+
+##### Enabling
+On enabling, Evacuation Coordinator:
+* preserves settings in the local file;
+* evacuation Coordinator enables local Eviction Agent;
+* passes to the evicting state.
+
+The internal API is:
+```erlang
+ok = emqx_node_rebalance_evacuation:start(
+    #{
+        conn_evict_rate => 10,
+        server_reference => <<"srv:123">>
```

Comment:
Q: since the server reference can be a list of servers, maybe this could be a list as well?

Reply:
MQTT doesn't specify exact Server-Reference format, so finally, I decided not to complexify things and treat it as an opaque value.

Reply:
From the spec, I think that it should be a space-separated list of servers.  Each server format is unspecified, but space-separated: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901255

> The Server Reference is a UTF-8 Encoded String. The value of this string is a space separated list of references. The format of references is not specified here.